domain/credential/service/service.go

```diff
@@ -124,3 +197,189 @@ func (s *Service) WatchCredential(ctx context.Context, tag names.CloudCredential
+	models, err := s.st.ModelsForCloudCredential(ctx, id)
+	if err != nil && !errors.Is(err, errors.NotFound) {
+		return nil, errors.Trace(err)
+	}
+	return models, nil
+}
+
+func (s *Service) validateCredentialForModel(modelUUID string, cld cloud.Cloud, id credential.ID, cred *cloud.Credential) []error {
+	ctx, err := s.validationContextGetter(modelUUID)
+	if err != nil {
+		return []error{errors.Trace(err)}
+	}
+	ctx.Cloud = cld
+	validator := s.validator
+
+	modelErrors, err := validator.Validate(ctx, id, cred, false)
+	if err != nil {
+		return []error{errors.Trace(err)}
+	}
+	return modelErrors
+}
+
+// UpdateCredentialModelResult holds details of a model
+// which was affected by a credential update, and any
+// errors encountered validating the credential.
+type UpdateCredentialModelResult struct {
+	// ModelUUID contains model's UUID.
+	ModelUUID string
+
+	// ModelName contains model name.
+	ModelName string
+
+	// Errors contains the errors accumulated while trying to update a credential.
+	Errors []error
+}
+
+// CheckAndUpdateCredential updates the credential after first checking that any models which use the credential
+// can still access the cloud resources. If force is true, update the credential even if there are issues
+// validating the credential.
+// TODO(wallyworld) - we need a strategy to handle changes which occur after the affected models have been read
+// but before validation can complete.
+func (s *Service) CheckAndUpdateCredential(ctx context.Context, id credential.ID, cred cloud.Credential, force bool) ([]UpdateCredentialModelResult, error) {
+	if s.validationContextGetter == nil {
+		return nil, errors.New("cannot validate credential with nil context getter")
+	}
+
+	models, err := s.credentialModels(ctx, id)
+	if err != nil && !force {
+		return nil, errors.Trace(err)
+	}
+
+	modelCloud, err := s.st.GetCloud(ctx, id.Cloud)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var (
+		modelsErred  bool
+		modelsResult []UpdateCredentialModelResult
+	)
+	if len(models) > 0 {
+		for uuid, name := range models {
+			result := UpdateCredentialModelResult{
+				ModelUUID: uuid,
+				ModelName: name,
+			}
+			result.Errors = s.validateCredentialForModel(uuid, modelCloud, id, &cred)
+			modelsResult = append(modelsResult, result)
+			if len(result.Errors) > 0 {
+				modelsErred = true
+			}
+		}
+		// Since we get a map above, for consistency ensure that models are added
+		// sorted by model uuid.
+		sort.Slice(modelsResult, func(i, j int) bool {
+			return modelsResult[i].ModelUUID < modelsResult[j].ModelUUID
+		})
+	}
+
+	if modelsErred && !force {
+		return modelsResult, credentialerrors.CredentialValidation
+	}
+
+	existingInvalid, err := s.st.UpsertCloudCredential(ctx, id.Name, id.Cloud, id.Owner, cred)
+	if err != nil {
+		if errors.Is(err, errors.NotFound) {
+			err = errors.Errorf(
+				"cannot update credential %q: controller does not manage cloud %q",
```

Comment:
Seems like if we are going to make assertions that a cloud or credential might live on another controller this type of error string should be getting produced at a user facing component.

The services layer doesn't even understand what it is talking about so just return a good typed error and let something else with a larger world view make larger claims.

Reply:
This check is simply that the cloud name in the credential id exists on the controller or not. It's a mistake to try and add a credential with the ID specifies cloud "bar" but the controller only has cloud "baz". It has nothing to do with if the credential exists or not - it's simply verifying that the cloud component of the credential  ID makes sense. This logic is pre-existing and I don't want to change the semantics of  credential ids in this PR. This logic is literally copied verbatim from the apiserver layer to get the organisation roughly correct. We can discuss re-modelling how credential IDs work later.

Reply:
Why? I prefer errors which tell you what went wrong. Juju suffers badly from generic not found errors being returned. Surely it would be better to introduce a new error type? The logic and error message is what we have now already in the apiserver layer, but we could have a new `InvalidControllerCloud` error.

Reply:
Yeah I am very ok with that.